src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -73,18 +81,178 @@ public void registered(SchedulerDriver driver, FrameworkID frameworkId, MasterIn
+    reconcileTasks(driver);
+  }
+
+  @Override
+  public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
+    log.info(String.format(
+        "Received status update for taskId=%s state=%s message='%s' stagingTasks.size=%d",
+        status.getTaskId().getValue(),
+        status.getState().toString(),
+        status.getMessage(),
+        liveState.getStagingTasksSize()));
+
+    if (!isStagingState(status)) {
+      liveState.removeStagingTask(status.getTaskId());
+    }
+
+    if (isTerminalState(status)) {
+      liveState.removeRunningTask(status.getTaskId());
+      persistentState.removeTaskId(status.getTaskId().getValue());
+      if (reconciliationComplete()) {
+        correctCurrentPhase();
+      }
+
+    } else if (isRunningState(status)) {
+      liveState.updateTaskForStatus(status);
+
+      log.info(String.format("Current Acquisition Phase: %s", liveState
+          .getCurrentAcquisitionPhase().toString()));
+
+      switch (liveState.getCurrentAcquisitionPhase()) {
+        case RECONCILING_TASKS :
+          if (reconciliationComplete()) {
+            correctCurrentPhase();
+          }
+          break;
+        case JOURNAL_NODES :
+          if (liveState.getJournalNodeSize() == conf.getJournalNodeCount()) {
+            correctCurrentPhase();
+          }
+          break;
+        case START_NAME_NODES :
+          if (liveState.getNameNodeSize() == (HDFSConstants.TOTAL_NAME_NODES)) {
+            // TODO move the reload to correctCurrentPhase and make it idempotent
+            reloadConfigsOnAllRunningTasks(driver);
+            correctCurrentPhase();
+          }
+          break;
+        case FORMAT_NAME_NODES :
+          if (!liveState.isNameNode1Initialized()) {
+            sendMessageTo(
+                driver,
+                liveState.getFirstNameNodeTaskId(),
+                liveState.getFirstNameNodeSlaveId(),
+                HDFSConstants.NAME_NODE_INIT_MESSAGE);
+          } else if (!liveState.isNameNode2Initialized()) {
+            sendMessageTo(
+                driver,
+                liveState.getSecondNameNodeTaskId(),
+                liveState.getSecondNameNodeSlaveId(),
+                HDFSConstants.NAME_NODE_BOOTSTRAP_MESSAGE);
+          } else {
+            correctCurrentPhase();
+          }
+          break;
+        // TODO (elingg) add a configurable number of data nodes
+        case DATA_NODES :
+          break;
+      }
+    } else {
+      log.warn(String.format("Don't know how to handle state=%s for taskId=%s",
+          status.getState(), status.getTaskId().getValue()));
+    }
+  }
+
+  @Override
+  public void resourceOffers(SchedulerDriver driver, List<Offer> offers) {
+    log.info(String.format("Received %d offers", offers.size()));
+
+    if (liveState.getStagingTasksSize() != 0) {
+      log.info("Declining offers because tasks are currently staging");
+      for (Offer offer : offers) {
+        driver.declineOffer(offer.getId());
```

Comment:
Related to the below comment about launching multiple tasks on one set of offers, you don't need to wait for staging tasks to launch more JNs, NNs, or DNs if that's the state you're in. For DNs especially, there's no need to wait for staging tasks.